### PR TITLE
Move Kibana configuration in a Secret

### LINKED ISSUE #946: Move Kibana configuration in a Secret
To expose the Kibana configuration to users (#740) we should start to move the `kibana.yml` configuration file in a `Secret`.

### PR DESCRIPTION
**Pull request by thbkrkr**
This commit moves the Kibana configuration directory in a Secret mounted as a Volume in each pod of the Kibana deployment.

- Migrate all env vars in the config settings file
- Reconcile a new config secret using a static `kibana.yml` config settings file (named `<name>-kb-config`)
- Update the deployment spec with the new config volume and volumeMount
- Reuse Elasticsearch settings CanonicalConfig bits (move in a `common` package)

Fixes #946.

### CODE REVIEW

File: operators/pkg/controller/common/settings/canonical_config.go [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+}
+
+// NewCanonicalConfigFrom creates a new config from the API type.
+func NewCanonicalConfigFrom(data untypedDict) (*CanonicalConfig, error) {
+	config, err := ucfg.NewFrom(data, Options...)
+	if err != nil {
+		return nil, err
+	}
+	return fromConfig(config), nil
+}
```
Comment (thbkrkr) [inline operators/pkg/controller/common/settings/canonical_config.go]: I updated the argument of `NewCanonicalConfigFrom` to remove the dependency to `estype.Config`.
https://github.com/elastic/cloud-on-k8s/blob/f00602603757aed48c15da96c557b40edb9a4dbe/operators/pkg/controller/elasticsearch/settings/canonical_config.go#L33-L39

File: operators/pkg/controller/common/settings/canonical_config.go [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+	return nil
+}
+
+// Unpack returns a typed config given a struct pointer.
+func (c *CanonicalConfig) Unpack(cfg interface{}) error {
+	if reflect.ValueOf(cfg).Kind() != reflect.Ptr {
+		panic("Unpack expects a struct pointer as argument")
+	}
+	return c.asUCfg().Unpack(cfg, Options...)
+}
```
Comment (thbkrkr) [inline operators/pkg/controller/common/settings/canonical_config.go]: I updated the return of `Unpack` to remove the dependency to `estype.ElasticsearchSettings`.
https://github.com/elastic/cloud-on-k8s/blob/f00602603757aed48c15da96c557b40edb9a4dbe/operators/pkg/controller/elasticsearch/settings/canonical_config.go#L91-L95

File: operators/pkg/controller/elasticsearch/settings/canonical_config.go [hunk 3]
```diff
@@ -4,284 +4,8 @@
...
-	return (*ucfg.Config)(c)
-}
-
-func fromConfig(in *ucfg.Config) *CanonicalConfig {
-	return (*CanonicalConfig)(in)
-}
+/*func Unpack(c *settings.CanonicalConfig) (v1alpha1.ElasticsearchSettings, error) {
+	cfg := v1alpha1.DefaultCfg
+	return cfg, c.AsUCfg().Unpack(&cfg, settings.Options...)
+}*/
```
Comment (pebrc) [inline operators/pkg/controller/elasticsearch/settings/canonical_config.go]: This should go away

File: operators/pkg/controller/elasticsearch/version/common.go [hunk 4]
```diff
@@ -265,10 +266,13 @@ func NewPod(
 	if objectMeta.Labels == nil {
 		objectMeta.Labels = map[string]string{}
 	}
-	cfg, err := podSpecCtx.Config.Unpack()
+
+	cfg := v1alpha1.DefaultCfg
+	err := podSpecCtx.Config.Unpack(&cfg)
```
Comment (pebrc) [inline operators/pkg/controller/elasticsearch/version/common.go]: I wonder if we could do something clever with a `type` around the shared config implementation to get back to a typesafe `Unpack()` specific to Elasticsearch without the need for developers to remember to use the defaulted struct.

```golang

// in: controllers/elasticsearch/settings
package settings
type CanonicalConfig struct {
 commonsettings.CanonicalConfig
}

func(c *CanonicalConfig) Unpack() (v1alpha1.ElasticsearchSettings, error) {
	cfg := v1alpha1.DefaultCfg
	return c.Unpack(cfg)
}
```

File: operators/pkg/controller/kibana/config/reconciler.go [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+			reconciled.Data = expected.Data
+		},
+	}); err != nil {
+		return err
+	}
+	return nil
+}
+
+// getConfigSecret returns the Kibana config secret for the given Kibana resource.
+func getConfigSecret(client k8s.Client, kb v1alpha1.Kibana) (*corev1.Secret, error) {
```
Comment (pebrc) [inline operators/pkg/controller/kibana/config/reconciler.go]: Naming: should this be called auth secret? Config secret implies for me the secret through which we mount the config files. But this seems to contain the Es auth info?

File: operators/pkg/controller/kibana/pod/pod.go [hunk 6]
```diff
@@ -32,15 +32,15 @@ func ApplyToEnv(auth v1alpha1.ElasticsearchAuth, env []corev1.EnvVar) []corev1.E
...
+			corev1.EnvVar{Name: elasticsearchPasswordEnvVar, Value: auth.Inline.Password},
 		)
 	}
 	if auth.SecretKeyRef != nil {
 		env = append(
 			env,
-			corev1.EnvVar{Name: elasticsearchUsername, Value: auth.SecretKeyRef.Key},
-			corev1.EnvVar{Name: elasticsearchPassword, ValueFrom: &corev1.EnvVarSource{
+			corev1.EnvVar{Name: elasticsearchUsernameEnvVar, Value: auth.SecretKeyRef.Key},
+			corev1.EnvVar{Name: elasticsearchPasswordEnvVar, ValueFrom: &corev1.EnvVarSource{
```
Comment (pebrc) [inline operators/pkg/controller/kibana/pod/pod.go]: This is unused code IIUC and should be removed

File: operators/pkg/controller/elasticsearch/mutation/comparison/config.go [hunk 7]
```diff
@@ -7,10 +7,11 @@ package comparison
 import (
 	"fmt"
 
+	commonsettings "github.com/elastic/cloud-on-k8s/operators/pkg/controller/common/settings"
```
Comment (pebrc) [inline operators/pkg/controller/elasticsearch/mutation/comparison/config.go]: alias is a bit long. No sure if I know a better alternative though 🤷‍♂ Maybe just `common`. BTW I don't like our `common` package name very much. It is as bad as `util`, `base` and friends. But that is a different question unrelated to this PR.

Comment (thbkrkr) [reply operators/pkg/controller/elasticsearch/mutation/comparison/config.go]: I hesitate... yes, a bit long. Go for `common`.

File: operators/pkg/controller/elasticsearch/settings/canonical_config.go (hunk 3 above)
Comment (thbkrkr) [reply operators/pkg/controller/elasticsearch/settings/canonical_config.go]: Oops.

File: operators/pkg/controller/kibana/config/reconciler.go (hunk 5 above)
Comment (thbkrkr) [reply operators/pkg/controller/kibana/config/reconciler.go]: Absolutely. I'll rename it.

File: operators/pkg/controller/elasticsearch/version/common.go (hunk 4 above)
Comment (thbkrkr) [reply operators/pkg/controller/elasticsearch/version/common.go]: _Yes, we can._

**Pull request merged**